### PR TITLE
Migrate to Pydantic v2

### PR DESCRIPTION
**Pull request by grahamalama**
I'd suggest reviewing this PR by commit. Tried to keep things as atomic as possible. At a high level, this PR:

- updates usages of renamed methods and attributes
- updates usages of deprecated methods and attributes
- Installs and configures `pydantic-settings`, which is now a separate package
- Configures new validators (at the field and model level) 

These commits warrant careful review:
- [Serialize timestamps consistently](https://github.com/mozilla-it/ctms-api/pull/942/commits/6704893e59f61f8fb80623875a988900e5b30bbe)
- [Coerce URLs to strings](https://github.com/mozilla-it/ctms-api/pull/942/commits/26870f855d39b060f9e67289e959792380ee1ba1)
- [Build new objects with factories and use "correct" schemas in backport tests](https://github.com/mozilla-it/ctms-api/pull/942/commits/2ddb58da834b065b9dec9d9777b54534dd15fc95)

### CODE REVIEW

File: ctms/config.py [hunk 1]
```diff
@@ -48,18 +51,10 @@ class Settings(BaseSettings):
     logging_level: LogLevel = LogLevel.INFO
     sentry_debug: bool = False
 
-    fastapi_env: Optional[str] = None
-    host: str = "0.0.0.0"
-    port: int = 8000
+    fastapi_env: Optional[str] = Field(default=None, alias="FASTAPI_ENV")
+    host: str = Field(default="0.0.0.0", alias="HOST")
+    port: int = Field(default=8000, alias="PORT")
```
Comment (grahamalama) [inline ctms/config.py:56]: These aliases provide a way to specify these settings without the `ctms_` prefix present on the other settings

File: tests/integration/test_basket_waitlist_subscription.py [hunk 2]
```diff
@@ -18,9 +18,9 @@ class Settings(BaseSettings):
     # We initialize CTMS api client id/secret in `ctms-db-init.sql`
     ctms_client_id: str = "id_integration-test"
     ctms_client_secret: str
-
-    class Config:
-        env_file = os.path.join(TEST_FOLDER, "basket.env")
+    model_config = SettingsConfigDict(
+        env_file=os.path.join(TEST_FOLDER, "basket.env"), extra="ignore"
```
Comment (grahamalama) [inline tests/integration/test_basket_waitlist_subscription.py:22]: Without `extra="ignore"`, any environment variable not specified in these settings triggers a validation error

**Review by alexcottner — APPROVED** [review]
Good work, this was a big effort for a dependency upgrade.
Don't see anything concerning that stands out.

File: ctms/schemas/contact.py [hunk 3]
```diff
@@ -192,15 +198,40 @@ class ContactPatchSchema(ComparableBase):
...
-    amo: Optional[Union[Literal["DELETE"], AddOnsInSchema]]
-    email: Optional[EmailPatchSchema]
-    fxa: Optional[Union[Literal["DELETE"], FirefoxAccountsInSchema]]
-    mofo: Optional[Union[Literal["DELETE"], MozillaFoundationInSchema]]
-    newsletters: Optional[Union[List[NewsletterSchema], Literal["UNSUBSCRIBE"]]]
-    waitlists: Optional[Union[List[WaitlistInSchema], Literal["UNSUBSCRIBE"]]]
-    # TODO waitlist: remove once Basket leverages the `waitlists` field.
-    vpn_waitlist: Optional[Union[Literal["DELETE"], VpnWaitlistInSchema]]
-    relay_waitlist: Optional[Union[Literal["DELETE"], RelayWaitlistInSchema]]
+    amo: Optional[Union[Literal["DELETE"], AddOnsInSchema]] = Field(
```
Comment (robhudson) [inline ctms/schemas/contact.py:210]: I'm new-ish to typing, but I'm curious why not the newer syntax of:
```python
    amo: Literal["DELETE"] | AddOnsInSchema | None = Field(
```

Comment (grahamalama) [reply ctms/schemas/contact.py:210]: Yeah I think that'd work too! For this PR, I was focused on doing as little as possible (which was still _a lot_ to make things v2 compatible and not touch much else).

I think doing things like refining these types would be good for a future PR.

**Review by leplatrem — APPROVED** [review]
👌

**Pull request merged**